go/libraries/doltcore/remotestorage/chunk_store.go

```diff
@@ -202,44 +203,123 @@ const (
+		for j < len(gr.Ranges) {
+			if gr.GapBetween(i, j) > maxGapBytes {
+				break
+			}
+			j++
+		}
+		res = append(res, &GetRange{Url: gr.Url, Ranges: gr.Ranges[i:j]})
+		i = j
+	}
+	return res
+}
+
+func (gr *GetRange) NumChunks() int {
+	return len(gr.Ranges)
+}
+
+func (gr *GetRange) RangeLen() uint64 {
+	return gr.ChunkEndOffset(gr.NumChunks()-1) - gr.ChunkStartOffset(0)
+}
+
+func (gr *GetRange) ChunkByteRange(i int) (uint64, uint64) {
+	start := gr.ChunkStartOffset(i) - gr.ChunkStartOffset(0)
+	end := gr.ChunkEndOffset(i) - gr.ChunkStartOffset(0)
+	return start, end
+}
+
+func (gr *GetRange) WorkFunc(ctx context.Context, fetcher HTTPFetcher, chunkChan chan nbs.CompressedChunk) func() error {
+	if len(gr.Ranges) == 0 {
+		return func() error { return nil }
+	}
+	return func() error {
+		comprData, err := rangeDownloadWithRetries(ctx, fetcher, gr.ChunkStartOffset(0), gr.RangeLen(), gr.Url)
+		if err != nil {
+			return err
+		}
+		// Send the chunk for each range included in GetRange.
+		for i := 0; i < len(gr.Ranges); i++ {
+			s, e := gr.ChunkByteRange(i)
+			cmpChnk, err := nbs.NewCompressedChunk(hash.New(gr.Ranges[i].Hash), comprData[s:e])
+			if err != nil {
+				return err
+			}
+			select {
+			case chunkChan <- cmpChnk:
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}
+		return nil
+	}
+}
+
+func (dcs *DoltChunkStore) getDLLocs(ctx context.Context, hashes []hash.Hash) (map[string]*GetRange, error) {
+	res := make(map[string]*GetRange)
 
 	// channel for receiving results from go routines making grpc calls to get download locations for chunks
-	dlLocChan := make(chan *remotesapi.DownloadLoc, len(hashes))
+	dlLocChan := make(chan []*remotesapi.HttpGetRange)
+
+	eg, ctx := errgroup.WithContext(ctx)
 
 	// go routine for receiving the results of the grpc calls and aggregating the results into resourceToUrlAndRanges
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		for loc := range dlLocChan {
-			switch typedLoc := loc.Location.(type) {
-			case *remotesapi.DownloadLoc_HttpGet:
-				panic("deprecated")
-			case *remotesapi.DownloadLoc_HttpGetRange:
-				if len(typedLoc.HttpGetRange.Ranges) > 0 {
-					urlStr := typedLoc.HttpGetRange.Url
-					urlObj, _ := url.Parse(urlStr)
-
-					resourcePath := fmt.Sprintf("%s://%s%s", urlObj.Scheme, urlObj.Host, urlObj.Path)
-
-					if uAndR, ok := resourceToUrlAndRanges[resourcePath]; ok {
-						uAndR.Ranges = append(uAndR.Ranges, typedLoc.HttpGetRange.Ranges...)
-						resourceToUrlAndRanges[resourcePath] = uAndR
+	eg.Go(func() error {
+		for {
+			select {
+			case locs, ok := <-dlLocChan:
+				if !ok {
+					return nil
+				}
+				for _, loc := range locs {
+					gr := (*GetRange)(loc)
+					if v, ok := res[gr.ResourcePath()]; ok {
+						v.Append(gr)
 					} else {
-						resourceToUrlAndRanges[resourcePath] = urlAndRanges{urlStr, typedLoc.HttpGetRange.Ranges}
+						res[gr.ResourcePath()] = gr
 					}
 				}
+			case <-ctx.Done():
+				return ctx.Err()
```

Comment:
I would return nil here.  This go routine did not error.  This won't affect anything, I just feel like its more correct.

Reply:
In these cases, I think it's best to still return `ctx.Err()`, because this could be the first goroutine that is encountering the canceled context. The cancelation could have been of a parent context, as opposed to it being of the `errgroup`'s context after the `errgroup` had already observed a non-nil error from one of its dispatched functions. This function was unable to fulfill its contract, which included reading `dlLocChan` to completion, so it returns an error about what went wrong...

Reply:
sounds good.